src/shim.js

```diff
@@ -149,6 +184,118 @@ function executeCommand(command) {
 }
 
 /**
+ * Adds an event to the target
+ *
+ * @param target the id of the object to listen
+ * @param eventName the name off the event
+ * @param callback the callback of the event
+ * @param args the additional arguments for the callback
+ */
+function addEvent(target, eventName, callback, args) {
+    let type = getTargetType(target);
+
+    if (isEventSupported(type, eventName)) {
+        let listeners = getEventListeners(target, eventName);
+
+        if (typeof callback === 'function') {
+            listeners.otherListeners.push(function () {
+                let params = [].slice.call(arguments).concat(args);
+                return callback.apply(objectSpace[target], params);
+            });
+        }
+    }
+}
+
+/**
+ * Removes all the listeners for an event in a target
+ *
+ * @param target the id of the target
+ * @param eventName the name of the event
+ */
+function removeEvent(target, eventName) {
+    let type = getTargetType(target);
+
+    if (isEventSupported(type, eventName)) {
+        events[target][eventName] = null;
+        objectSpace[target][eventName] = null;
+    }
+}
+
+/**
+ * Verifies if an event is supported for a type of target
+ *
+ * @param type
+ * @param eventName
+ * @returns {boolean}
+ */
+function isEventSupported(type, eventName) {
+    return eventsForEntities[type] && eventsForEntities[type].indexOf(eventName) !== -1;
+}
+
+/**
+ * Gets an object containing all the listeners for an event of a target
+ *
+ * @param target the target id
+ * @param eventName the event name
+ */
+function getEventListeners(target, eventName) {
+    if (!events[target]) {
+        events[target] = {};
+    }
+
+    if (!events[target][eventName]) {
+        events[target][eventName] = {
+            outsideListener: getOutsideListener(eventName, target),
+            otherListeners: []
+        };
+
+        objectSpace[target][eventName] = triggerEvent.bind(null, target, eventName);
+    }
+
+    return events[target][eventName];
+}
+
+/**
+ * Determines a targets type using its id
+ *
+ * @param target
+ * @returns {*}
+ */
+function getTargetType(target) {
+    return target.toString().split('$')[0];
+}
+
+/**
+ * Executes all the listeners for an event from a target
+ *
+ * @param target
+ * @param eventName
+ */
+function triggerEvent(target, eventName) {
+    let args = [].slice.call(arguments, 2);
+    let listeners = events[target][eventName];
+    listeners.outsideListener.apply(null, args);
+    listeners.otherListeners.forEach(function (listener) {
+        listener.apply(objectSpace[target], args);
+    });
+}
+
+/**
+ * Returns a function that will notify to node that an event have been triggered
+ *
+ * @param eventName
+ * @param targetId
+ * @returns {Function}
+ */
+function getOutsideListener(eventName, targetId) {
+    return function () {
+        var args = [].slice.call(arguments, 0);
+        system.stdout.writeLine('<event>' + JSON.stringify({target: targetId, type: eventName, args}));
```

Comment:
Are you sure? I have a feeling if you comment out the heartbeat, your events will never trigger. Because `let line = system.stdin.readLine();` is a blocking call. So when it finds something to parse, it will process it and write back to node. Then it will go back to `stdin`. Since it is blocking and phantomjs is single threaded, it will sit there until a heart beat is triggered. At that point, it will see an event in phantomjs and write something to nodejs. 

Overall though, nice job. Clean code and I like it a lot. But I believe this does require the heartbeat to let it run.